crates/red_knot_python_semantic/src/types.rs

```diff
@@ -521,6 +521,53 @@ impl<'db> Type<'db> {
         }
     }
 
+    /// Resolves the boolean value of a type.
+    /// This is used to determine the value that would be returned when `__bool__` is called on an object
+    pub fn bool(&self, db: &'db dyn Db) -> Option<bool> {
+        match self {
+            Type::Any => None,
+            Type::Never => None,
+            Type::Unknown => None,
+            Type::Unbound => None,
+            Type::None => Some(false),
+            Type::Function(_) | Type::RevealTypeFunction(_) => Some(true),
+            Type::Module(_) => Some(true),
+            Type::Class(_) => Some(true),
```

Comment:
Not necessarily! Most classes are always truthy, but not if the class has a custom metaclass that defines `__bool__` or `__len__`:

```pycon
>>> class Meta(type):
...     def __bool__(cls): return False
... 
>>> class Foo(metaclass=Meta): pass
... 
>>> bool(Foo)
False
```

The exact details around the method lookup will be somewhat fiddly (for both instances and classes -- you need to check both `__bool__` and `__len__`), so I think it's fine to just return `None` for both this variant and `Type::Instance` for now. But we should add a TODO so that we remember to add some logic doing the method lookup for these variants at some point. (If a type has a `__bool__` method that is annotated as returning `Literal[False]`, or a `__len__` method annotated as returning `Literal[0]`, we should treat it as always falsey -- and similarly for `__bool__` and `__len__` methods that are annotated as returning `Literal[True]` / `Literal[<some nonzero number>]`.)